File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/ElectronicHorizonParams.java
package com.mapbox.services.android.navigation.v5.navigation;

public class ElectronicHorizonParams {

  private long delay;
  private long interval;
  private int locationsCacheSize;

  private ElectronicHorizonParams(long delay, long interval, int locationsCacheSize) {
    this.delay = delay;
    this.interval = interval;
    this.locationsCacheSize = locationsCacheSize;
  }

  long getDelay() {
    return delay;
  }

  long getInterval() {
    return interval;
  }

  int getLocationsCacheSize() {
    return locationsCacheSize;
  }

  public static class Builder {
    private static final long ELECTRONIC_HORIZON_DELAY_DEFAULT = 20_000;
    private static final long ELECTRONIC_HORIZON_INTERVAL_DEFAULT = 20_000;
    private static final int LOCATIONS_CACHE_SIZE_DEFAULT = 5;
    private static final int LOCATIONS_CACHE_MIN_SIZE = 2;
    private static final int LOCATIONS_CACHE_MAX_SIZE = 10;

    private long delay = ELECTRONIC_HORIZON_DELAY_DEFAULT;
    private long interval = ELECTRONIC_HORIZON_INTERVAL_DEFAULT;
    private int locationsCacheSize = LOCATIONS_CACHE_SIZE_DEFAULT;

    public Builder delay(long delay) {
      if (delay > 0) {
        this.delay = delay;
      }
      return this;
    }

    public Builder interval(long interval) {
      if (interval > 0) {
        this.interval = interval;
      }
      return this;
    }

    public Builder locationsCacheSize(int locationsCacheSize) {
      if (locationsCacheSize >= LOCATIONS_CACHE_MIN_SIZE && locationsCacheSize <= LOCATIONS_CACHE_MAX_SIZE) {
        this.locationsCacheSize = locationsCacheSize;
      }
      return this;
    }

    public ElectronicHorizonParams build() {
      return new ElectronicHorizonParams(delay, interval, locationsCacheSize);
    }
  }
}

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/freedrive/ElectronicHorizonParams.kt
package com.mapbox.navigation.core.freedrive

class ElectronicHorizonParams private constructor(
        val delay: Long,
        val interval: Long,
        val locationsCacheSize: Int
) {
    class Builder {
        companion object {
            internal const val DEFAULT_ELECTRONIC_HORIZON_DELAY = 20_000L
            internal const val DEFAULT_ELECTRONIC_HORIZON_INTERVAL = 20_000L
            internal const val DEFAULT_LOCATIONS_CACHE_SIZE = 5
            internal const val LOCATIONS_CACHE_MIN_SIZE = 2
            internal const val LOCATIONS_CACHE_MAX_SIZE = 10
        }

        private var delay: Long = DEFAULT_ELECTRONIC_HORIZON_DELAY
        private var interval: Long = DEFAULT_ELECTRONIC_HORIZON_INTERVAL
        private var locationsCacheSize: Int = DEFAULT_LOCATIONS_CACHE_SIZE

        fun delay(delay: Long) = apply {
            invokeIf(delay > 0) { this.delay = delay }
        }

        fun interval(interval: Long) = apply {
            invokeIf(interval > 0) { this.interval = interval }
        }

        fun locationsCacheSize(cacheSize: Int) = apply {
            invokeIf(cacheSize in LOCATIONS_CACHE_MIN_SIZE..LOCATIONS_CACHE_MAX_SIZE) {
                this.locationsCacheSize = cacheSize
            }
        }

        fun build() = ElectronicHorizonParams(
                delay,
                interval,
                locationsCacheSize
        )

        private inline fun invokeIf(condition: Boolean, block: () -> Unit) {
            if (condition) {
                block()
            }
        }
    }
}